File: rd-cpp/src/rd_core_cpp/src/main/util/Logger.cpp
#include "Logger.h"

#include "core_util.h"

#include <iostream>
#include <thread>

/*SwitchLogger::SwitchLogger(const std::string &category) {}

SwitchLogger get_logger(std::string category) {
    return SwitchLogger(category);
}*/

/*void SwitchLogger::log(LogLevel level, std::string message, std::exception const &e) {
    realLogger.log(level, message, e);
}

bool SwitchLogger::is_enabled(LogLevel level) {
    return realLogger.isEnabled(level);
}*/

namespace rd {
	//namespace log {
	LogLevel minimum_level_to_log;
	//}
	using namespace std::literals::string_literals;

	std::string to_string(LogLevel level) {
		switch (level) {
			case LogLevel::Trace:
				return "Trace";
			case LogLevel::Debug:
				return "Debug";
			case LogLevel::Info:
				return "Info ";
			case LogLevel::Warn:
				return "Warn ";
			case LogLevel::Error:
				return "Error";
			case LogLevel::Fatal:
				return "Fatal";
		}
		return {};
	}

	std::string string_format(string_view format, va_list args)
	{
		size_t len = std::vsnprintf(NULL, 0, format.data(), args);
		va_end(args);
		std::string vec;
		vec.resize(len + 1);
//		va_start(args, format);
		std::vsnprintf(&vec[0], len + 1, format.data(), args);
		va_end(args);
		return vec;
	}

	void Logger::log(LogLevel level, string_view format, va_list args, std::exception const *e) const {
		if (level >= minimum_level_to_log) {
			std::stringstream ss;
			ss << to_string(level) << " | " << to_string(std::this_thread::get_id()) << " | ";
			ss << string_format(format, args);
			if (e) {
				ss << " | " << e->what();
			}
			std::cerr << ss.str() << std::endl;
		}
	}

	void Logger::trace(const std::exception *e, string_view msg, ...) const {
		va_list args;
		va_start(args, msg);
		log(LogLevel::Error, msg, args, e);
	}

	void Logger::trace(string_view msg, ...) const {
		va_list args;
		va_start(args, msg);
		log(LogLevel::Error, msg, args);
	}

	void Logger::debug(const std::exception *e, string_view msg, ...) const {
		va_list args;
		va_start(args, msg);
		log(LogLevel::Error, msg, args, e);
	}

	void Logger::debug(string_view msg, ...) const {
		va_list args;
		va_start(args, msg);
		log(LogLevel::Error, msg, args);
	}

	void Logger::info(const std::exception *e, string_view msg, ...) const {
		va_list args;
		va_start(args, msg);
		log(LogLevel::Error, msg, args, e);
	}

	void Logger::info(string_view msg, ...) const {
		va_list args;
		va_start(args, msg);
		log(LogLevel::Error, msg, args);
	}

	void Logger::warn(const std::exception *e, string_view msg, ...) const {
		va_list args;
		va_start(args, msg);
		log(LogLevel::Error, msg, args, e);
	}

	void Logger::warn(string_view msg, ...) const {
		va_list args;
		va_start(args, msg);
		log(LogLevel::Error, msg, args);
	}

	void Logger::error(std::exception const *e, string_view msg, ...) const {
		va_list args;
		va_start(args, msg);
		log(LogLevel::Error, msg, args, e);
	}

	void Logger::error(string_view msg, ...) const {
		va_list args;
		va_start(args, msg);
		log(LogLevel::Error, msg, args);
	}
}

File: rd-kt/rd-core/src/commonMain/kotlin/com/jetbrains/rd/util/Logger.kt
package com.jetbrains.rd.util

import com.jetbrains.rd.util.lifetime.Lifetime
import com.jetbrains.rd.util.lifetime.onTermination
import com.jetbrains.rd.util.reactive.viewableTail
import kotlin.reflect.KClass

enum class LogLevel {
    Trace,
    Debug,
    Info,
    Warn,
    Error,
    Fatal
}

interface Logger {
    companion object {
        val root = getLogger("")

        fun set(lifetime: Lifetime, logger: ILoggerFactory) {
            lifetime.onTermination ( //careful - this executes instantly and returns disposable which is added by onTermination
                Statics<ILoggerFactory>().push(logger)
            )
        }
    }
    fun log(level: LogLevel, message: Any?, throwable: Throwable?)
    fun isEnabled(level: LogLevel): Boolean
}

interface ILoggerFactory {
    fun getLogger(category: String): Logger
}


class SwitchLogger(category: String) : Logger {
    private lateinit var realLogger: Logger

    init {
        val stack = Statics<ILoggerFactory>().stack
        stack.viewableTail().advise(Lifetime.Eternal) { factory ->
            realLogger = (factory?:ConsoleLoggerFactory).getLogger(category)
        }
    }

    override fun log(level: LogLevel, message: Any?, throwable: Throwable?) {
        realLogger.log(level, message, throwable)
    }

    override fun isEnabled(level: LogLevel): Boolean = realLogger.isEnabled(level)

}


object ConsoleLoggerFactory : ILoggerFactory {
    var minLevelToLog : LogLevel = LogLevel.Debug
    var levelToLogStderr : LogLevel? = LogLevel.Warn

    var traceCategories = mutableSetOf<String>()

    override fun getLogger(category: String) = object : Logger {

        override fun log(level: LogLevel, message: Any?, throwable: Throwable?) {
            if (!isEnabled(level)) return

            val msg = defaultLogFormat(category, level, message, throwable)
            if (levelToLogStderr?.let { level > it } == true)
                printlnError(msg)
            else
                println(msg)
        }

        override fun isEnabled(level: LogLevel): Boolean = level >=  this@ConsoleLoggerFactory.minLevelToLog ||
            traceCategories.contains(category) || traceCategories.contains(category.substringBefore('.'))
    }
}

//This could be changed via Statics or similar plugin technique
fun defaultLogFormat(category: String, level: LogLevel, message: Any?, throwable: Throwable?) : String {
    val throwableToPrint = if (level < LogLevel.Error) throwable  else throwable ?: Exception() //to print stacktrace
    return "${level.toString().padEnd(5)} | ${category.substringAfterLast('.').padEnd(25)} | ${currentThreadName().padEnd(15)} | ${message?.toString()?:""} ${throwableToPrint?.getThrowableText()?.let { "| $it" }?:""}"
}

fun getLogger(category: String) = SwitchLogger(category)
fun getLogger(categoryKclass: KClass<*>): Logger = getLogger(qualifiedName(categoryKclass))
inline fun <reified T> getLogger() = getLogger(T::class)


inline fun Logger.log(level: LogLevel, msg: () -> Any?) {
    if (isEnabled(level)) {
        log(level, msg(), null)
    }
}

inline fun Logger.trace(msg: () -> Any?) = log(LogLevel.Trace, msg)
inline fun Logger.debug(msg: () -> Any?) = log(LogLevel.Debug, msg)
inline fun Logger.info(msg: () -> Any?)  = log(LogLevel.Info, msg)
inline fun Logger.warn(msg: () -> Any?)  = log(LogLevel.Warn, msg)
inline fun Logger.error(msg: () -> Any?) = log(LogLevel.Error, msg)

fun Logger.error(msg: String?, throwable: Throwable) = log(LogLevel.Error, msg, throwable)
fun Logger.error(throwable: Throwable) = this.error(null, throwable)

inline fun Logger.catch(comment: String?, action:() -> Unit) {
    try {
        action()
    } catch (e : Throwable) {
        val sfx = if (comment.isNullOrBlank()) "" else ": $comment"
        error("Catch$sfx", e)
    }
}
inline fun Logger.catch(action:() -> Unit) = catch (null, action)

inline fun catchAndDrop(action:() -> Unit) {
    try {
        action()
    } catch (e : Throwable) {
        //nothing
    }
}

inline fun catch(comment: String?, action:() -> Unit) {
    try {
        action()
    } catch (e : Throwable) {
        val sfx = if (comment.isNullOrBlank()) "" else ": $comment"
        getLogger("Default-Error-Logger").log(LogLevel.Error, "Catch$sfx", e)
    }
}
inline fun catch(action:() -> Unit) = catch (null, action)
